File: firmware/CarTelemetry/LiveTelemetry.m
%% ================================================================
%  LiveTelemetry.m — Real-time car telemetry dashboard
%
%  Reads LoRa receiver serial data and plots live subplots for
%  speed, temperature, acceleration, gyroscope, RSSI, and SNR.
%
%  LoRa receiver format:
%    +RCV=<addr>,<len>,<payload>,<RSSI>,<SNR>
%
%  Payload format (pipe-delimited):
%    pkt|aX|aY|aZ|gX|gY|gZ|spdMph|objF|forceN|lat|lon|sats
%
%  Usage:
%    LiveTelemetry("COM5")            % Windows
%    LiveTelemetry("/dev/ttyUSB0")    % Linux / Mac
%    LiveTelemetry("COM5", 300)       % custom window of 300 samples
% =================================================================

function LiveTelemetry(comPort, maxPoints)

    arguments
        comPort   string
        maxPoints double = 200   % rolling window size
    end

    BAUD = 115200;

    % -- Field indices inside the pipe-delimited payload -----------
    F_PKT   = 1;
    F_AX    = 2;   F_AY  = 3;   F_AZ  = 4;
    F_GX    = 5;   F_GY  = 6;   F_GZ  = 7;
    F_SPEED = 8;
    F_TEMP  = 9;
    F_FORCE = 10;
    F_LAT   = 11;  F_LON = 12;
    F_SATS  = 13;

    % -- Preallocate circular buffers --------------------------------
    N       = maxPoints;
    time_s  = NaN(1, N);      % elapsed seconds
    ax_buf  = NaN(1, N);  ay_buf  = NaN(1, N);  az_buf  = NaN(1, N);
    gx_buf  = NaN(1, N);  gy_buf  = NaN(1, N);  gz_buf  = NaN(1, N);
    spd_buf = NaN(1, N);
    tmp_buf = NaN(1, N);
    frc_buf = NaN(1, N);
    rssi_buf= NaN(1, N);
    snr_buf = NaN(1, N);
    sat_buf = NaN(1, N);
    lat_buf = NaN(1, N);  lon_buf = NaN(1, N);
    idx     = 0;

    % -- Open serial port --------------------------------------------
    fprintf("Opening %s at %d baud...\n", comPort, BAUD);
    try
        s = serialport(comPort, BAUD);
        configureTerminator(s, "LF");
        s.Timeout = 10;
    catch err
        error("Could not open %s: %s", comPort, err.message);
    end
    cleanupObj = onCleanup(@() delete(s));   % auto-close on exit

    fprintf("Connected. Waiting for data...\n\n");

    % -- Build figure & subplots -------------------------------------
    fig = figure('Name', 'Car Telemetry — Live', ...
                 'NumberTitle', 'off', ...
                 'Color', [0.12 0.12 0.14], ...
                 'Position', [50 50 1400 900]);

    tileset = tiledlayout(fig, 4, 3, 'TileSpacing', 'compact', ...
                          'Padding', 'compact');
    title(tileset, 'Live Car Telemetry', ...
          'Color', 'w', 'FontSize', 16, 'FontWeight', 'bold');

    % Acceleration X / Y / Z
    axAx = nexttile(tileset);  hAx = animatedline(axAx, 'Color', [0.25 0.80 0.90], 'LineWidth', 1.2);
    stylePlot(axAx, 'Accel X', 'g');

    axAy = nexttile(tileset);  hAy = animatedline(axAy, 'Color', [0.40 0.85 0.50], 'LineWidth', 1.2);
    stylePlot(axAy, 'Accel Y', 'g');

    axAz = nexttile(tileset);  hAz = animatedline(axAz, 'Color', [0.95 0.55 0.25], 'LineWidth', 1.2);
    stylePlot(axAz, 'Accel Z', 'g');

    % Gyroscope X / Y / Z
    axGx = nexttile(tileset);  hGx = animatedline(axGx, 'Color', [0.25 0.80 0.90], 'LineWidth', 1.2);
    stylePlot(axGx, 'Gyro X', '°/s');

    axGy = nexttile(tileset);  hGy = animatedline(axGy, 'Color', [0.40 0.85 0.50], 'LineWidth', 1.2);
    stylePlot(axGy, 'Gyro Y', '°/s');

    axGz = nexttile(tileset);  hGz = animatedline(axGz, 'Color', [0.95 0.55 0.25], 'LineWidth', 1.2);
    stylePlot(axGz, 'Gyro Z', '°/s');

    % Speed / Temp / Force
    axSpd = nexttile(tileset);  hSpd = animatedline(axSpd, 'Color', [0.95 0.30 0.35], 'LineWidth', 1.4);
    stylePlot(axSpd, 'Speed', 'mph');

    axTmp = nexttile(tileset);  hTmp = animatedline(axTmp, 'Color', [1.0 0.75 0.20], 'LineWidth', 1.4);
    stylePlot(axTmp, 'Object Temp', '°F');

    axFrc = nexttile(tileset);  hFrc = animatedline(axFrc, 'Color', [0.70 0.45 0.95], 'LineWidth', 1.4);
    stylePlot(axFrc, 'Force', 'N');

    % RSSI / SNR / Satellites
    axRssi = nexttile(tileset);  hRssi = animatedline(axRssi, 'Color', [0.90 0.35 0.55], 'LineWidth', 1.2);
    stylePlot(axRssi, 'RSSI', 'dBm');

    axSnr = nexttile(tileset);  hSnr = animatedline(axSnr, 'Color', [0.35 0.75 0.95], 'LineWidth', 1.2);
    stylePlot(axSnr, 'SNR', 'dB');

    axSat = nexttile(tileset);  hSat = animatedline(axSat, 'Color', [0.50 0.90 0.40], 'LineWidth', 1.2);
    stylePlot(axSat, 'Satellites', '#');

    % -- Status text (lat/lon, packet #) -----------------------------
    statusTxt = annotation(fig, 'textbox', [0.01 0.005 0.98 0.03], ...
        'String', 'Waiting for first packet...', ...
        'Color', [0.7 0.7 0.7], 'EdgeColor', 'none', ...
        'FontSize', 10, 'FontName', 'Consolas', ...
        'HorizontalAlignment', 'center', ...
        'FitBoxToText', 'off');

    startTic = tic;

    % -- Main read loop ----------------------------------------------
    while isvalid(fig)
        try
            line = readline(s);
        catch
            continue;   % timeout — just retry
        end

        if strlength(line) == 0, continue; end

        % Parse +RCV=<addr>,<len>,<payload>,<RSSI>,<SNR>
        [payload, rssi, snr, ok] = parseRCV(line);
        if ~ok, continue; end

        % Split pipe-delimited payload
        fields = split(payload, '|');
        if numel(fields) < F_SATS, continue; end

        % -- Extract values ------------------------------------------
        elapsed = toc(startTic);
        pkt     = str2double(fields{F_PKT});
        aX      = str2double(fields{F_AX});
        aY      = str2double(fields{F_AY});
        aZ      = str2double(fields{F_AZ});
        gX      = str2double(fields{F_GX});
        gY      = str2double(fields{F_GY});
        gZ      = str2double(fields{F_GZ});
        speed   = str2double(fields{F_SPEED});
        objTemp = str2double(fields{F_TEMP});
        forceN  = str2double(fields{F_FORCE});
        lat     = str2double(fields{F_LAT});
        lon     = str2double(fields{F_LON});
        sats    = str2double(fields{F_SATS});

        % -- Console output ------------------------------------------
        fprintf(['[#%04d]  Accel: %+.2f %+.2f %+.2f g  |  ' ...
                 'Gyro: %+.1f %+.1f %+.1f  |  ' ...
                 'Spd: %.1f mph  |  Temp: %.1fF  |  ' ...
                 'Force: %.1f N  |  ' ...
                 'RSSI: %d  SNR: %d  |  ' ...
                 'Sats: %d  (%.6f, %.6f)\n'], ...
                pkt, aX, aY, aZ, gX, gY, gZ, ...
                speed, objTemp, forceN, rssi, snr, sats, lat, lon);

        % -- Push into animated lines ---------------------------------
        addpoints(hAx,  elapsed, aX);
        addpoints(hAy,  elapsed, aY);
        addpoints(hAz,  elapsed, aZ);
        addpoints(hGx,  elapsed, gX);
        addpoints(hGy,  elapsed, gY);
        addpoints(hGz,  elapsed, gZ);
        addpoints(hSpd, elapsed, speed);
        addpoints(hTmp, elapsed, objTemp);
        addpoints(hFrc, elapsed, forceN);
        addpoints(hRssi,elapsed, rssi);
        addpoints(hSnr, elapsed, snr);
        addpoints(hSat, elapsed, sats);

        % Slide x-axis window
        if elapsed > 60
            windowStart = elapsed - 60;
            setAxLimits([axAx axAy axAz axGx axGy axGz ...
                         axSpd axTmp axFrc axRssi axSnr axSat], ...
                        windowStart, elapsed);
        end

        % -- Update status bar ----------------------------------------
        statusTxt.String = sprintf( ...
            'Pkt #%d  |  Lat: %.6f  Lon: %.6f  |  Sats: %d  |  RSSI: %d dBm  |  SNR: %d dB  |  Elapsed: %.0fs', ...
            pkt, lat, lon, sats, rssi, snr, elapsed);

        drawnow limitrate;
    end

    fprintf("\nFigure closed — stopping.\n");
end

%% ================================================================
%  Parse a +RCV line from the LoRa receiver
%  Format: +RCV=<addr>,<len>,<payload>,<RSSI>,<SNR>
% =================================================================
function [payload, rssi, snr, ok] = parseRCV(line)
    payload = ""; rssi = 0; snr = 0; ok = false;

    line = strip(line);
    if ~startsWith(line, "+RCV="), return; end

    % Strip the "+RCV=" prefix
    body = extractAfter(line, "+RCV=");

    % Split on comma — but payload itself has no commas (pipe-delimited)
    parts = split(body, ',');
    % Expected: addr, len, payload, RSSI, SNR  (5 parts)
    if numel(parts) < 5, return; end

    payload = parts{3};
    rssi    = str2double(parts{end-1});
    snr     = str2double(parts{end});
    ok      = true;
end

%% ================================================================
%  Apply consistent dark styling to a subplot
% =================================================================
function stylePlot(ax, titleStr, unitStr)
    title(ax, titleStr, 'Color', 'w', 'FontSize', 11);
    ylabel(ax, unitStr, 'Color', [0.7 0.7 0.7]);
    ax.Color           = [0.15 0.15 0.18];
    ax.XColor          = [0.5 0.5 0.5];
    ax.YColor          = [0.5 0.5 0.5];
    ax.GridColor       = [0.3 0.3 0.3];
    ax.GridAlpha       = 0.6;
    ax.XGrid           = 'on';
    ax.YGrid           = 'on';
    ax.Box             = 'on';
    ax.FontSize        = 9;
end

%% ================================================================
%  Slide x-axis limits for all axes
% =================================================================
function setAxLimits(axes_array, xmin, xmax)
    for i = 1:numel(axes_array)
        axes_array(i).XLim = [xmin, xmax];
    end
end

File: firmware/CarTelemetry/LiveTelemetry.py
"""
LiveTelemetry.py — Real-time car telemetry dashboard

Reads LoRa receiver serial data and plots live subplots for
speed, temperature, acceleration, gyroscope, RSSI, and SNR.

Requirements:
    pip install pyserial matplotlib

Usage:
    python LiveTelemetry.py COM4            # Windows
    python LiveTelemetry.py /dev/ttyUSB0    # Linux/Mac
    python LiveTelemetry.py COM4 300        # custom rolling window
"""

import sys
import serial
import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
import matplotlib.animation as animation
from collections import deque
import time

# ── Config ─────────────────────────────────────────────────────
BAUD = 115200
WINDOW = 200           # default rolling window size (samples)

# Payload field indices (pipe-delimited)
# pkt|aX|aY|aZ|gX|gY|gZ|spdMph|objF|forceN|lat|lon|sats
F_PKT, F_AX, F_AY, F_AZ = 0, 1, 2, 3
F_GX, F_GY, F_GZ = 4, 5, 6
F_SPEED, F_TEMP, F_FORCE = 7, 8, 9
F_LAT, F_LON, F_SATS = 10, 11, 12
MIN_FIELDS = 13


def parse_rcv(line):
    """Parse +RCV=<addr>,<len>,<payload>,<RSSI>,<SNR>"""
    line = line.strip()
    if not line.startswith("+RCV="):
        return None

    body = line[5:]          # strip "+RCV="
    parts = body.split(",")
    if len(parts) < 5:
        return None

    payload = parts[2]
    rssi = int(parts[-2])
    snr = int(parts[-1])
    return payload, rssi, snr


def main():
    if len(sys.argv) < 2:
        print("Usage: python LiveTelemetry.py <COM_PORT> [window_size]")
        print("  e.g. python LiveTelemetry.py COM4")
        sys.exit(1)

    port = sys.argv[1]
    window = int(sys.argv[2]) if len(sys.argv) > 2 else WINDOW

    # ── Open serial ────────────────────────────────────────────
    print(f"Opening {port} at {BAUD} baud...")
    try:
        ser = serial.Serial(port, BAUD, timeout=2)
    except serial.SerialException as e:
        print(f"Error: {e}")
        sys.exit(1)
    print("Connected. Waiting for data...\n")

    # ── Buffers ────────────────────────────────────────────────
    t_buf    = deque(maxlen=window)
    ax_buf   = deque(maxlen=window)
    ay_buf   = deque(maxlen=window)
    az_buf   = deque(maxlen=window)
    gx_buf   = deque(maxlen=window)
    gy_buf   = deque(maxlen=window)
    gz_buf   = deque(maxlen=window)
    spd_buf  = deque(maxlen=window)
    tmp_buf  = deque(maxlen=window)
    frc_buf  = deque(maxlen=window)
    rssi_buf = deque(maxlen=window)
    snr_buf  = deque(maxlen=window)
    sat_buf  = deque(maxlen=window)

    start_time = time.time()

    # ── Figure setup ───────────────────────────────────────────
    plt.style.use('dark_background')
    fig, axes = plt.subplots(4, 3, figsize=(16, 10))
    fig.suptitle('Live Car Telemetry', fontsize=16, fontweight='bold')
    fig.subplots_adjust(hspace=0.45, wspace=0.30)

    plot_cfg = [
        # (row, col, buffer, label, unit, color)
        (0, 0, ax_buf,   'Accel X',     'g',    '#40CCE6'),
        (0, 1, ay_buf,   'Accel Y',     'g',    '#66D980'),
        (0, 2, az_buf,   'Accel Z',     'g',    '#F28C40'),
        (1, 0, gx_buf,   'Gyro X',      '°/s',  '#40CCE6'),
        (1, 1, gy_buf,   'Gyro Y',      '°/s',  '#66D980'),
        (1, 2, gz_buf,   'Gyro Z',      '°/s',  '#F28C40'),
        (2, 0, spd_buf,  'Speed',       'mph',  '#F24D59'),
        (2, 1, tmp_buf,  'Object Temp', '°F',   '#FFC033'),
        (2, 2, frc_buf,  'Force',       'N',    '#B373F2'),
        (3, 0, rssi_buf, 'RSSI',        'dBm',  '#E65989'),
        (3, 1, snr_buf,  'SNR',         'dB',   '#59BFF2'),
        (3, 2, sat_buf,  'Satellites',  '#',    '#80E666'),
    ]

    lines = []
    for r, c, _, label, unit, color in plot_cfg:
        ax = axes[r][c]
        ln, = ax.plot([], [], color=color, linewidth=1.2)
        ax.set_title(label, fontsize=11, color='white')
        ax.set_ylabel(unit, fontsize=9, color='#B0B0B0')
        ax.tick_params(colors='#808080', labelsize=8)
        ax.grid(True, alpha=0.3)
        lines.append(ln)

    status_text = fig.text(0.50, 0.01, 'Waiting for first packet...',
                           ha='center', fontsize=10, color='#B0B0B0',
                           family='monospace')

    # ── Animation update ───────────────────────────────────────
    def update(frame):
        try:
            raw = ser.readline().decode('utf-8', errors='ignore')
        except Exception:
            return lines

        result = parse_rcv(raw)
        if result is None:
            return lines

        payload, rssi, snr = result
        fields = payload.split('|')
        if len(fields) < MIN_FIELDS:
            return lines

        try:
            elapsed = time.time() - start_time
            pkt   = int(fields[F_PKT])
            aX    = float(fields[F_AX])
            aY    = float(fields[F_AY])
            aZ    = float(fields[F_AZ])
            gX    = float(fields[F_GX])
            gY    = float(fields[F_GY])
            gZ    = float(fields[F_GZ])
            speed = float(fields[F_SPEED])
            temp  = float(fields[F_TEMP])
            force = float(fields[F_FORCE])
            lat   = float(fields[F_LAT])
            lon   = float(fields[F_LON])
            sats  = int(fields[F_SATS])
        except (ValueError, IndexError):
            return lines

        # Console output
        print(f"[#{pkt:04d}]  Accel: {aX:+.2f} {aY:+.2f} {aZ:+.2f} g  |  "
              f"Gyro: {gX:+.1f} {gY:+.1f} {gZ:+.1f}  |  "
              f"Spd: {speed:.1f} mph  |  Temp: {temp:.1f}F  |  "
              f"Force: {force:.1f} N  |  "
              f"RSSI: {rssi}  SNR: {snr}  |  "
              f"Sats: {sats}  ({lat:.6f}, {lon:.6f})")

        # Push to buffers
        t_buf.append(elapsed)
        ax_buf.append(aX);    ay_buf.append(aY);    az_buf.append(aZ)
        gx_buf.append(gX);    gy_buf.append(gY);    gz_buf.append(gZ)
        spd_buf.append(speed); tmp_buf.append(temp); frc_buf.append(force)
        rssi_buf.append(rssi); snr_buf.append(snr); sat_buf.append(sats)

        # Update plot lines
        all_bufs = [ax_buf, ay_buf, az_buf,
                    gx_buf, gy_buf, gz_buf,
                    spd_buf, tmp_buf, frc_buf,
                    rssi_buf, snr_buf, sat_buf]

        for i, (r, c, buf, _, _, _) in enumerate(plot_cfg):
            lines[i].set_data(list(t_buf), list(buf))
            ax = axes[r][c]
            if len(t_buf) > 1:
                ax.set_xlim(t_buf[0], t_buf[-1])
            if len(buf) > 0:
                bmin, bmax = min(buf), max(buf)
                margin = max(abs(bmax - bmin) * 0.15, 0.5)
                ax.set_ylim(bmin - margin, bmax + margin)

        status_text.set_text(
            f'Pkt #{pkt}  |  Lat: {lat:.6f}  Lon: {lon:.6f}  |  '
            f'Sats: {sats}  |  RSSI: {rssi} dBm  |  SNR: {snr} dB  |  '
            f'Elapsed: {elapsed:.0f}s')

        return lines

    ani = animation.FuncAnimation(fig, update, interval=50, blit=False,
                                  cache_frame_data=False)
    plt.show()

    ser.close()
    print("\nWindow closed — stopped.")


if __name__ == "__main__":
    main()
